File: app/src/main/java/com/worldventures/dreamtrips/social/ui/friends/presenter/FriendPreferencesPresenter.java
package com.worldventures.dreamtrips.social.ui.friends.presenter;

import com.innahema.collections.query.queriables.Queryable;
import com.worldventures.core.janet.CommandWithError;
import com.worldventures.core.model.Circle;
import com.worldventures.core.model.User;
import com.worldventures.dreamtrips.modules.common.presenter.Presenter;
import com.worldventures.dreamtrips.modules.common.view.BlockingProgressView;
import com.worldventures.dreamtrips.social.service.friends.interactor.CirclesInteractor;
import com.worldventures.dreamtrips.social.service.friends.interactor.command.GetCirclesCommand;
import com.worldventures.dreamtrips.social.ui.profile.bundle.UserBundle;
import com.worldventures.dreamtrips.social.ui.profile.model.FriendGroupRelation;
import com.worldventures.dreamtrips.social.ui.profile.service.ProfileInteractor;
import com.worldventures.dreamtrips.social.ui.profile.service.command.AddFriendToCircleCommand;
import com.worldventures.dreamtrips.social.ui.profile.service.command.RemoveFriendFromCircleCommand;
import com.worldventures.dreamtrips.social.ui.profile.view.cell.delegate.State;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.techery.janet.helper.ActionStateSubscriber;
import rx.android.schedulers.AndroidSchedulers;

public class FriendPreferencesPresenter extends Presenter<FriendPreferencesPresenter.View> {

   @Inject CirclesInteractor circlesInteractor;
   @Inject ProfileInteractor profileInteractor;

   private User friend;

   public FriendPreferencesPresenter(UserBundle userBundle) {
      friend = userBundle.getUser();
   }

   @Override
   public void takeView(View view) {
      super.takeView(view);
      subscribeCircles();
      updateCircles();
   }

   private void updateCircles() {
      circlesInteractor.getPipe().send(new GetCirclesCommand());
   }

   private void subscribeCircles() {
      circlesInteractor.getPipe()
            .observe()
            .observeOn(AndroidSchedulers.mainThread())
            .compose(bindView())
            .subscribe(new ActionStateSubscriber<GetCirclesCommand>().onStart(circlesCommand -> onCirclesStart())
                  .onSuccess(circlesCommand -> onCirclesSuccess(new ArrayList<>(circlesCommand.getResult())))
                  .onFail(this::onCirclesError));
   }

   private void onCirclesStart() {
      view.showBlockingProgress();
   }

   private void onCirclesSuccess(List<Circle> resultCircles) {
      view.addItems(Queryable.from(resultCircles).map(circle -> {
         return new FriendGroupRelation(circle, friend);
      }).toList());
      view.hideBlockingProgress();
   }

   private void onCirclesError(CommandWithError commandWithError, Throwable throwable) {
      view.hideBlockingProgress();
      handleError(commandWithError, throwable);
   }

   public void onRelationshipChanged(Circle circle, State state) {
      switch (state) {
         case ADDED:
            profileInteractor.addFriendToCirclesPipe().send(new AddFriendToCircleCommand(circle, friend));
            friend.getCircles().add(circle);
            break;
         case REMOVED:
            profileInteractor.removeFriendFromCirclesPipe().send(new RemoveFriendFromCircleCommand(circle, friend));
            friend.getCircles().remove(circle);
            break;
         default:
            break;
      }

   }

   public interface View extends Presenter.View, BlockingProgressView {
      void addItems(List<FriendGroupRelation> circles);
   }
}

File: app/src/main/java/com/worldventures/dreamtrips/social/ui/friends/presenter/FriendPreferencesPresenter.kt
package com.worldventures.dreamtrips.social.ui.friends.presenter

import com.worldventures.core.model.Circle
import com.worldventures.core.model.User
import com.worldventures.dreamtrips.modules.common.presenter.Presenter
import com.worldventures.dreamtrips.modules.common.view.BlockingProgressView
import com.worldventures.dreamtrips.social.service.profile.ProfileInteractor
import com.worldventures.dreamtrips.social.service.profile.command.AddFriendToCircleCommand
import com.worldventures.dreamtrips.social.service.profile.command.RemoveFriendFromCircleCommand
import com.worldventures.dreamtrips.social.service.profile.model.FriendGroupRelation
import com.worldventures.dreamtrips.social.service.users.base.interactor.CirclesInteractor
import com.worldventures.dreamtrips.social.service.users.circle.command.GetCirclesCommand
import com.worldventures.dreamtrips.social.ui.profile.bundle.UserBundle
import com.worldventures.dreamtrips.social.ui.profile.view.cell.delegate.State
import io.techery.janet.helper.ActionStateSubscriber
import rx.android.schedulers.AndroidSchedulers
import javax.inject.Inject

class FriendPreferencesPresenter(userBundle: UserBundle) : Presenter<FriendPreferencesPresenter.View>() {

   @Inject lateinit var circlesInteractor: CirclesInteractor
   @Inject lateinit var profileInteractor: ProfileInteractor

   val friend: User = userBundle.user

   override fun takeView(view: View?) {
      super.takeView(view)
      subscribeCircles()
      updateCircles()
   }

   private fun subscribeCircles() {
      circlesInteractor.pipe
            .observe()
            .observeOn(AndroidSchedulers.mainThread())
            .compose(bindViewToMainComposer())
            .subscribe(ActionStateSubscriber<GetCirclesCommand>()
                  .onStart { view.showBlockingProgress() }
                  .onSuccess {
                     view.addItems(it.result.map { FriendGroupRelation(it, friend) }.toList())
                     view.hideBlockingProgress()
                  }
                  .onFail({ command, error ->
                     view.hideBlockingProgress()
                     handleError(command, error)
                  }))
   }

   private fun updateCircles() {
      circlesInteractor.pipe.send(GetCirclesCommand())
   }

   fun onRelationshipChanged(circle: Circle, state: State) {
      when (state) {
         State.ADDED -> {
            profileInteractor.addFriendToCirclePipe.send(AddFriendToCircleCommand(circle, friend))
            friend.circles.add(circle)
         }
         State.REMOVED -> {
            profileInteractor.removeFriendFromCirclePipe
                  .send(RemoveFriendFromCircleCommand(circle, friend))
         }
      }
   }

   interface View : Presenter.View, BlockingProgressView {
      fun addItems(circles: List<FriendGroupRelation>)
   }
}
